processor/swok8sworkloadtypeprocessor/processor.go

```diff
@@ -0,0 +1,224 @@
+	"fmt"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.opentelemetry.io/collector/processor"
+	"go.uber.org/zap"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+type swok8sworkloadtypeProcessor struct {
+	logger    *zap.Logger
+	cancel    context.CancelFunc
+	config    *Config
+	settings  processor.Settings
+	k8sClient kubernetes.Interface
+	factory   informers.SharedInformerFactory
+	informers map[K8sWorkloadType]cache.SharedIndexInformer
+}
+
+func (cp *swok8sworkloadtypeProcessor) getAttribute(attributes pcommon.Map, attrName string) string {
+	if attrVal, ok := attributes.Get(attrName); ok && attrVal.Type() == pcommon.ValueTypeStr {
+		return attrVal.Str()
+	}
+	return ""
+}
+
+func (cp *swok8sworkloadtypeProcessor) setWorkloadTypeForGivenDatapointAttributes(attributes pcommon.Map) error {
+	for _, workloadMapping := range cp.config.WorkloadMappings {
+		name := cp.getAttribute(attributes, workloadMapping.NameAttr)
+		if name == "" {
+			continue
+		}
+		namespace := cp.getAttribute(attributes, workloadMapping.NamespaceAttr)
+		var workloadKey string
+		if namespace != "" {
+			workloadKey = fmt.Sprintf("%s/%s", namespace, name)
+		} else {
+			workloadKey = name
+		}
+
+		for _, workloadType := range workloadMapping.ExpectedTypes {
+			workload, exists, err := cp.informers[workloadType].GetStore().GetByKey(workloadKey)
+			if err != nil {
+				return fmt.Errorf("error when retrieving workload '%s' from cache: %w", workloadKey, err)
+			}
+			if exists {
+				workloadObject, ok := workload.(runtime.Object)
+				if !ok {
+					return fmt.Errorf("unexpected type '%T' for workload '%s'", workload, workloadKey)
+				}
+				kind := workloadObject.GetObjectKind().GroupVersionKind().Kind
+				if kind == "" {
+					return fmt.Errorf("workload '%s' has no kind", workloadKey)
+				}
+				attributes.PutStr(workloadMapping.WorkloadTypeAttr, kind)
+				break
+			}
+		}
+	}
+
+	return nil
+}
+
+func (cp *swok8sworkloadtypeProcessor) processMetrics(ctx context.Context, md pmetric.Metrics) (pmetric.Metrics, error) {
+	resourceMetrics := md.ResourceMetrics()
+	for rmIdx := range resourceMetrics.Len() {
+		rm := resourceMetrics.At(rmIdx)
+		scopeMetrics := rm.ScopeMetrics()
+		for smIdx := range scopeMetrics.Len() {
+			sm := scopeMetrics.At(smIdx)
+			metrics := sm.Metrics()
+			for mIdx := range metrics.Len() {
+				m := metrics.At(mIdx)
+				switch m.Type() {
+				case pmetric.MetricTypeGauge:
+					datapoints := m.Gauge().DataPoints()
+					for dpIdx := range datapoints.Len() {
+						dp := datapoints.At(dpIdx)
+						err := cp.setWorkloadTypeForGivenDatapointAttributes(dp.Attributes())
+						if err != nil {
+							return pmetric.Metrics{}, fmt.Errorf("error getting workload type for metric '%s': %w", m.Name(), err)
```

Comment:
if an error occurs when processing a metric datapoint, the entire processing stops, and all subsequent metrics/data points are skipped. Shouldn't we just log error and continue processing to avoid loosing data?

Reply:
Yeah, I checked the contribution guidelines for the OTEL collector and they suggest to not return errors unless really necessary. So I changed the logic in this processor accordingly.